src/plugins/chart_expressions/expression_xy/public/helpers/data_layers.tsx

```diff
@@ -71,58 +71,85 @@ type GetColorFn = (
   }
 ) => string | null;
 
+export interface DatatableWithFormatInfo {
+  table: Datatable;
+  formattedColumns: Record<string, true>;
+}
+
+export type DatatablesWithFormatInfo = Record<string, DatatableWithFormatInfo>;
+
+export type FormattedDatatables = Record<string, Datatable>;
+
 const isPrimitive = (value: unknown): boolean => value != null && typeof value !== 'object';
 
-export const getFormattedTable = (
-  table: Datatable,
+export const getFormattedRow = (
+  row: Datatable['rows'][number],
+  columns: Datatable['columns'],
   formatFactory: FormatFactory,
   xAccessor: string | undefined,
   xScaleType: XScaleType
-): Datatable => ({
-  ...table,
-  rows: table.rows.map((row: DatatableRow) => {
-    const newRow = { ...row };
-    for (const column of table.columns) {
-      const record = newRow[column.id];
+): { row: Datatable['rows'][number]; formattedColumns: Record<string, true> } =>
+  columns.reduce(
+    (formattedInfo, { id, meta }) => {
+      const record = formattedInfo.row[id];
       if (
         record != null &&
         // pre-format values for ordinal x axes because there can only be a single x axis formatter on chart level
-        (!isPrimitive(record) || (column.id === xAccessor && xScaleType === 'ordinal'))
+        (!isPrimitive(record) || (id === xAccessor && xScaleType === 'ordinal'))
       ) {
-        newRow[column.id] = formatFactory(column.meta.params)!.convert(record);
+        return {
+          row: { ...formattedInfo.row, [id]: formatFactory(meta.params)!.convert(record) },
```

Comment:
Constructing the formatter via `formatFactory` is pretty expensive if executed once per row. As the formatter will be identical for each row, we should lift it out so it's only called once for each column

Reply:
Done.